### DISCUSSION TITLE
Expanding upon null-state static analysis and null state attributes to supply other exhaustiveness checks and accidently solve DUs along the way

### DISCUSSION THREAD

**Discussion by DavidArno** [dotnet/csharplang#5538] [Language Ideas]
C#'s null-state analysis, backed up by the nullable-analysis attributes enables better exhaustiveness checking in switches:
```c#
#nullable enable
using System.Diagnostics.CodeAnalysis;

[return: NotNull] static string? F() => "";

_ = F() switch {
    { Length: _ } => 1
};
```
In the above case, the compiler knows that the value from `F()` won't be `null`, so it doesn't require a null check in the switch. 
The `{ Length: _ }` pattern is exhaustive by itself.

But what if this were expanded to more generalized constrained values. For example:
```c#
enum TrafficLights { Red, Amber, Green }

[return: ConstrainedToNamedValues(typeof(TrafficLights))]
static TrafficLights F() => TrafficLights.Red;

_ = F() switch {
    TrafficLights.Red => 1,
    TrafficLights.Amber => 2,
    TrafficLights.Green => 3
};
```
Here, we've introduced a new static analysis attribute: `ConstrainedToNamedValues`, which in this context informs the compiler that only the named values of the enum will be returned by the function. This then allows the compiler to treat the switch as exhaustive.

As well as the enum-specific attribute above, we could then have a more general "anything in this set/range" constraint:
```c#
_ = F() switch {
    Apple a => ...,
    Banana b => ...,
    Durian => ...
};

[return:ConstrainedTo(typeof(Apple), typeof(Banana), typeof(Durian))] 
static object F() => new Durian();

record Apple(int Redness);
record Banana(double bendCoefficient);
record Durian();
```
Again, here the compiler is being told that the return value from `F()` is constrained to one of three types: `Apple`, `Banana` and `Durian`. So the switch can be treated as exhaustive.

At this point, some syntactic sugar could be introduced to method declarations:
```c#
static Apple | Banana | Durian F() => new Durian();
``` 
And when we define a discriminated union of `Fruit`:
```c#
enum struct Fruit
{
    Apple { int redness },
    Banana {  double bendCoefficient },
    Durian
}
```

Then the `Fruit` type could imply a property following this principle:
```c#
[return:ConstrainedTo(typeof(Apple), typeof(Banana), typeof(Durian))] Case => ...
``` 
Which in turn would allow for a closed collection of values that could be exhaustively patterned matched as well.

**Comment by HaloFour** [comment]
A chunk of this feels like it should be evaluated as a part of #113

**Reply by DavidArno** [reply]
The DU bit was actually an accidental afterthought. It started out just as a way of constraining enums for pattern matching but grew as I typed.

**Comment by CyrusNajmabadi** [comment]
I must reject this proposal due to the virtue that bananas are not fruits, they are berries.  We could not allow such a design that would allow for such a misclassification to be foisted on the world :(

**Reply by CyrusNajmabadi** [reply]
Note: the above is obviously a joke.  My true feelings are that this an interesting idea :)

**Reply by Richiban** [reply]
Surely a berry is a subtype of fruit?

**Reply by DavidArno** [reply]
@Richiban,

One shouldn't allow facts to get in the way of a good joke. Yes, all berries are fruits, so bananas, pumpkins, aubergines (eggplants) and tomatoes are all fruits due to being berries. But since that adds a layer of unwanted pedantry to Cyrus's comment, we turned a blind eye 😃

**Comment by 333fred** [comment]
It's an interesting proposal, thank for the tag on discord @DavidArno. In some ways, this feels like DUs via erasure: that object return being constrained with an attribute feels odd. But at the same time, I could _absolutely_ have used something like that when implementing the interpolated string handler feature in the compiler, as there were a number of places where I needed to operate on either a `BoundInterpolatedString` or a `BoundBinaryOperator`. Those two share a subtype of `BoundExpression`, and while I could introduce a new interface that both implement and work in terms of that, that sure is a lot of ceremony (and wouldn't be super useful in many parts of code that operate in terms of `BoundNode`s, not arbitrary interfaces), so I mostly ended up using `BoundExpression` as my interchange type and asserting where needed.

That being said, while doing union types via attributes and erasure is certainly an easy approach, I think the seams would start showing at the edges. Passing these through generics, for example, might drop constraints in unexpected ways. I do think if we want to pursue a generalized union type, I'd really like it to not be erased. Whether that's via creating enum structs and shipping a `OneOf` type with sugar over it, or modifying metadata to create a token that allows for multiple type tokens to be combined, I do think that approach will provide us with lasting benefits in the long run.

**Reply by bbarry** [reply]
If generics were variadic there could be a special generic type or two introduced to the language:

```csharp
static NamedValues<TrafficLights> F() => TrafficLights.Red;
static OneOf<Apple), Banana, Durian> Fruit() => new Durian();
```

It seems like either it needs to erase somehow or be tied to a new runtime version.

**Reply by DavidArno** [reply]
Yes, upon reflection I think that trying to extend the idea to full-blown DU types was a mistake. It takes the idea too far.

I still like the idea of it offering a way of doing lightweight union returns, `apple | orange F() => ...`. It almost feels like the union equivalent of what tuples offer compared with records. Each of tuples and records has it's place, so maybe lightweight unions using an idea like this have a place too.

Anyway, thanks for taking the time to read and feedback. Appreciated.
